src/charm.py

```diff
@@ -169,32 +242,79 @@ def _on_collect_status(self, event: CollectStatusEvent):
         if not self._is_vault_service_started():
             event.add_status(WaitingStatus("Waiting for Vault service to start"))
             return
+        if not self._api_address:
+            event.add_status(WaitingStatus("No address received from Juju yet"))
+            return
+        if not self.tls.tls_file_available_in_charm(File.CA):
+            event.add_status(WaitingStatus("Certificate is unavailable in the charm"))
+            return
+        vault = self._get_vault_client()
+        if not vault.is_api_available():
+            event.add_status(WaitingStatus("Vault API is not yet available"))
+            return
+        if not vault.is_initialized():
+            event.add_status(BlockedStatus("Waiting for Vault to be initialized"))
+            return
+        if vault.is_sealed():
+            event.add_status(BlockedStatus("Waiting for Vault to be unsealed"))
+            return
+        if not self._get_vault_approle_secret():
+            event.add_status(BlockedStatus("Waiting for charm to be authorized"))
+            return
         event.add_status(ActiveStatus())
 
-    def _configure(self, _):
+    def _configure(self, _):  # noqa: C901
         """Handle Vault installation.
 
         This includes:
           - Installing the Vault snap
           - Generating the Vault config file
         """
+        self._create_backend_directory()
+        self._create_certs_directory()
         self._install_vault_snap()
         if not self._is_peer_relation_created():
             return
         if not self._bind_address:
             return
-        if not self.unit.is_leader() and len(self._other_peer_node_api_addresses()) == 0:
-            return
-        if not self.unit.is_leader() and not self.tls.ca_certificate_is_saved():
-            return
-        self._create_backend_directory()
-        self._create_certs_directory()
+        if not self.unit.is_leader():
+            if len(self._other_peer_node_api_addresses()) == 0:
+                return
+            if not self.tls.ca_certificate_is_saved():
+                return
         self.tls.configure_certificates(self._bind_address)
         self._generate_vault_config_file()
         self._start_vault_service()
         self._set_peer_relation_node_api_address()
         self.tls.send_ca_cert()
 
+        if not self._api_address or not self.tls.tls_file_available_in_charm(File.CA):
+            return
+        vault = self._get_vault_client()
+        if not vault.is_api_available():
+            return
+        if not vault.is_initialized():
+            return
+        if vault.is_sealed():
+            return
+        if not (approle_auth := self._get_vault_approle_secret()):
+            return
+        vault.authenticate(AppRole(approle_auth[0], approle_auth[1]))
+
+        if vault.is_active() and not vault.is_raft_cluster_healthy():
+            logger.warning("Raft cluster is not healthy: %s", vault.get_raft_cluster_state())
```

Comment:
I'm not sure I understand the value of this block. What are we doing here?

Reply:
This is all the preconditions to validate if the cluster is healthy, just so we can log a warning. If the vault is still sealed, it will throw an error. And you need the API, vault to be initialized, and credentials to authenticate the vault client.

The reason behind logging this was a compromise for https://github.com/canonical/vault-k8s-operator/issues/100

Reply:
Should we move it to collect status so that if vault is unhealthy it is reflected in the charm status?

If we do decide to keep it here, should we move this block to its own method? 

Doing either of those will likely remove the need for the ` C901` on the `_configure`

Reply:
There was some discussion with @saltiyazan about whether we should reflect this in the charm status or not. I can't remember the exact reasoning, but he might have a better memory since he worked on it.

Regardless, if we don't like this approach I think we should break it out into a separate change and I can remove it from here for now. It is not required, only meant to provide a little extra information in the event something does go wrong. I was just trying to bring over what we already had in the k8s charm.

i'd be happy to move it to a separate method, though, if that's the preferred path for now.